### ISSUE TITLE
HTML for thumbnails: redundant "title"?

### ISSUE THREAD

**Issue by mgeier** [sphinx-gallery/sphinx-gallery#998]
I've copied the HTML generated by Sphinx Gallery for `nbsphinx` (https://github.com/spatialaudio/nbsphinx/pull/666) in order to solve #984.

If I'm not mistaken, the new HTML is something like this:

```html
<div class="sphx-glr-thumbcontainer" tooltip="{{ tooltip }}">
  <img alt="{{ title }}" src="{{ filename }}" />
  <p><a class="reference internal" href="{{ uri }}"><span class="std std-ref">{{ title }}</span></a></p>
  <div class="sphx-glr-thumbnail-title">{{ title }}</div>
</div>
```

It looks like the title appears three times.
Is that intentional?
Or am I missing something?

I think it is generally important to use good `alt` text for images, but in this case it seems to me that:

* it is entirely redundant
* the title is meant for the whole page, not necessarily for the specific thumbnail image

I'm not a user of a screen reader, but I have the feeling that the title might be spoken multiple times.

It might be good to use `alt=""`, which means that the image is not a key part of the content and AFAIU, it will not be spoken by a screen reader.

I don't really understand the `<span>` element within the `<a>` element, why is the title included there another time?

**Comment by larsoner** [comment]
For you perhaps @alexisthual ?

**Comment by StefRe** [comment]
If the title in the python file includes rst formatting, the titles are different:  

A python file with the title
```
"""
**bold** and `open`
===================
"""
```

will produce
```
<div class="sphx-glr-thumbcontainer" tooltip="**bold** and open">
  <img alt="**bold** and `open`" src="../_images/sphx_glr_rst_test_thumb.png" />
  <p><a class="reference internal" href="rst_test.html#sphx-glr-auto-examples-rst-test-py"><span class="std std-ref">bold and open</span></a></p>
  <div class="sphx-glr-thumbnail-title">**bold** and `open`</div>
</div>
```
which looks like this:
![image](https://user-images.githubusercontent.com/19879328/189999669-112982b7-a470-4466-9ae4-b0dc9c663ff9.png)

whereas the `sphx-glr-thumbnail-title` and the `tooltip` should be simply `bold and open` without asterisks and backticks just like in the `span`.

**Comment by mgeier** [comment]
Thanks @StefRe for the information!

So the `<span>` has its markup stripped, while the `<div>` still has the markdown.
Latter is likely a bug (https://github.com/sphinx-gallery/sphinx-gallery/issues/1002), but this still doesn't explain why the title is appearing multiple times and what the `<span>` is supposed to do.

Any ideas?

**Comment by StefRe** [comment]
I don't know enough of `docutils` to understand what's going on here, therefore I opened the regression issue. #906 replaced the `figure` directive by the `image` directive. The former includes a properly formatted caption whereas the latter lacks it so it was manually added as `sphx-glr-thumbnail-title` which isn't sanitized.

**Comment by lucyleeow** [comment]
> If I'm not mistaken, the new HTML is something like this:

Checked SG docs and get same format.

> I'm not a user of a screen reader, but I have the feeling that the title might be spoken multiple times.

Is there any way to check this? It would be a shame to remove it if wasn't spoken? 

FYI @alexisthual you'll need to amend associated tests if you are removing: https://github.com/sphinx-gallery/sphinx-gallery/pull/668/files

**Comment by mgeier** [comment]
I'm wondering why such complicated HTML is generated in the first place.

Why not generate something like this:

```html
<a class="reference internal sphx-glr-thumbcontainer" href="{{ uri }}" tooltip="{{ tooltip }}">
  <img alt="" src="{{ filename }}" />
  {{ title }}
</a>
```

AFAICT, this already looks as expected with the given CSS, and I think this would make some (a lot?) of the current CSS obsolete.

It would be much clearer that the whole thing is a link, and it would actually behave like a link, i.e. the color would change when visited.

And maybe a screen reader would handle it as expected, since it is just a normal `<a>` tag.

And it would fix my original issue: no redundant titles anymore!

----

@lucyleeow 

> Is there any way to check this?

Ideally, somebody who regularly uses a screen reader should check this.
If such a person is not available, anybody could install a screen reader, I guess. It's also possible to test with screen readers on BrowserStack, but I haven't really tried this.

> It would be a shame to remove it if wasn't spoken?

Yes, it would be a shame. But if the title is not spoken, something is broken anyway, and I don't think adding a redundant `alt` text would be a good solution.

**Comment by alexisthual** [comment]
> Why not generate something like this:
> 
> ```
> <a class="reference internal sphx-glr-thumbcontainer" href="{{ uri }}" tooltip="{{ tooltip }}">
>   <img alt="" src="{{ filename }}" />
>   {{ title }}
> </a>
> ```

I had tried two approaches in the past to get the HTML structure closer to what you suggest, but I wasn't able to do it easily. I had mostly tried two approaches:
- throwing custom HTML to a sphinx `:ref:`, but it doesn't seem doable
- implementing a custom `<a>` (to get rid of `:ref:`) whose `href` would be the one generated by `:ref:` but I couldn't find a method which yields the computed href of a given `:ref:`

I just did some more research though and came across an example in [`sphinx-panels`](https://sphinx-panels.readthedocs.io/en/latest/#) that makes use of `::after` to stretch a link so that it fills its parent `<div>`. I guess I could  implement something that looks like this:
```html
<div class="sphx-glr-thumbcontainer">
  <img src="{{filename}}" />
  <p><a class="reference internal" href="..." ><span>{{title}}</span></a></p>
  <div class="sphx-glr-thumbnail-desciption">{{tooltip}}</div>
</div>
```

Of course, if someone finds how to get a string with the href to the gallery item, we could simplify it further.

**Comment by alexisthual** [comment]
I had a closer look at the function which generates the thumbnail: https://github.com/sphinx-gallery/sphinx-gallery/blob/ff4886ace1c3d14ac4c3046572e5d62c56d4b16e/sphinx_gallery/backreferences.py#L286-L306

I think `target_dir` and `fname` should be enough to generate a valid href to the gallery item html file :slightly_smiling_face: I guess I can try to integrate this in #1005.

**Comment by alexisthual** [comment]
Actually it's not that simple to infer the href of a gallery item. @larsoner do you know of a function I could call to get this href?

**Comment by larsoner** [comment]
I *think* that by knowing the output/destination folder structure (e.g., `examples`->`auto_examples`) and input .py filenames you can construct it. I'm not sure we have a separate function for this

**Comment by mgeier** [comment]
FYI, I'm currently working on an alternative for `nbsphinx` at https://github.com/spatialaudio/nbsphinx/pull/706 which uses this HTML structure:

```html
<div class="nbsphinx-gallery">
<a class="reference internal" href="{uri}" title="{tooltip}">
  <div><img alt="" src="{filename}"></div>
  <div>{title}</div>
</a>
...
</div>
```

**Comment by larsoner** [comment]
Fixed by #1005
